nvram_plugin.py

```diff
@@ -0,0 +1,186 @@
+
+from pandare import PyPlugin, Panda
+import ipaddress
+import socket
+import os
+'''
+Interface to LibNvram Hypercalls.
+
+Register callbacks via 
+panda.pyplugins.ppp.NVramHC.ppp_reg_cb("nvram_get",get_ipdb_call)
+For NVRAM_GET exmaple
+TODO: Setup actual list add. Handle additional nvram hooks
+'''
+MAGIC_VAL = 0x31838188
+ERROR_VAL = 0x12345678
+class NVramHC(PyPlugin):
+    def __init__(self, panda: Panda):
+        print("Loading Nvram HyperCall plugin")
+        super().__init__(panda)
+        self.panda = panda
+        outdir = self.get_arg("outdir")
+        self.log = self.get_arg_bool("log")
+        if self.log:
+            self.write_file = os.path.join(outdir,"nvram.log")
+            open(self.write_file, "w").close() #Clear the file
+        self.nvram:dict = self.get_arg("nvram")
+        if self.nvram is None:
+            self.nvram = {}
+        self.create_callbacks()
+
+        self.conf = 0
+        @panda.cb_guest_hypercall
+        def hypercall(cpu):
+            if panda.arch.get_arg(cpu, 0, convention="syscall") != MAGIC_VAL:
+                print(f"Found hypercall, but value was {hex(panda.arch.get_arg(cpu, 0, convention='syscall'))}")
+                return False
+            len = panda.arch.get_arg(cpu, 1, convention='syscall')
+            pointer_size = panda.bits //8
+            argptr =  panda.arch.get_arg(cpu, 2, convention='syscall')
+            argstr = []
+            for x in range(0,len):
+                addr = int.from_bytes(panda.virtual_memory_read(cpu, argptr + ( pointer_size * x), pointer_size),'little')
+                argstr.append(panda.read_str(cpu, addr))
+            with open(self.write_file, "a") as f:
+                f.write(f"\n{argstr}\n")
+            block = self.ppp_run_cb("nvram_block_all",cpu,argstr)
+            if block:
+                return True
+            output = argstr[0]
+            if "NVRAM_GET_INT" in output:
```

Comment:
Unlikely, but an nvram key or value could contain these strings. If you encoded these as an int and in your argument to the hypercall you'd avoid that